crates/uv-resolver/src/resolution/display.rs

```diff
@@ -377,14 +331,45 @@ fn combine_extras(graph: &IntermediatePetGraph) -> RequirementsTxtGraph {
             continue;
         };
 
+        // In the `requirements.txt` output, we want a flat installation list, so we need to use
+        // the reachability markers instead of the edge markers.
+        // We use the markers of the base package: We know that each virtual extra package has an
+        // edge to the base package, so we know that base package markers are more general than the
+        // extra package markers (the extra package markers are a subset of the base package
+        // markers).
```

Comment:
This seems like an interesting insight that might be worth copying over to the docs on `PubGrubPackage`. Specifically, is it a guaranteed invariant that for any package `foo`, if there is both a `PubGrubPackage::Package` and a `PubGrubPackage::Marker` for it, then the former is always a superset of the latter?

(I'd love to start collecting invariants in the docs for `PubGrubPackage` because I find that type somewhat difficult to reason about. But I don't have the full shape in my head yet to propose something better.)

Reply:
I have a hard time documenting this because "the base package is always selected when one of its virtual packages is selected" seems so fundamental to me. Do we have a good place to put overall resolver things? Because I think it would be valuable to write some of them up, we bet we come across some improvements to uv by this way.

Reply:
I'm not sure that was always even true, was it? The way virtual packages get added has shifted around a lot I think. Particularly in `get_dependencies` in the resolver. I don't think anything fundamental has changed in ~months, but I know @charliermarsh did a lot of work there.

I would start by adding these sorts of things to `PubGrubPackage` I think. It is a very key data type in the resolver and I _think_ there are a lot of interesting invariants around it.